sklearn/tests/test_random_projection.py

```diff
@@ -334,23 +344,26 @@ def test_correct_RandomProjection_dimensions_embedding():
             assert 85 < rp.components_.nnz  # close to 1% density
 
 
-def test_warning_n_components_greater_than_n_features():
+@pytest.mark.parametrize("coo_container", COO_CONTAINERS)
+def test_warning_n_components_greater_than_n_features(coo_container):
     n_features = 20
-    data, _ = make_sparse_random_data(5, n_features, int(n_features / 4))
+    data, _ = make_sparse_random_data(coo_container, 5, n_features, int(n_features / 4))
 
     for RandomProjection in all_RandomProjection:
         with pytest.warns(DataDimensionalityWarning):
             RandomProjection(n_components=n_features + 1).fit(data)
 
 
-def test_works_with_sparse_data():
+@pytest.mark.parametrize("csr_container", CSR_CONTAINERS)
+@pytest.mark.parametrize("coo_container", COO_CONTAINERS)
+def test_works_with_sparse_data(csr_container, coo_container):
     n_features = 20
-    data, _ = make_sparse_random_data(5, n_features, int(n_features / 4))
+    data, _ = make_sparse_random_data(coo_container, 5, n_features, int(n_features / 4))
 
     for RandomProjection in all_RandomProjection:
         rp_dense = RandomProjection(n_components=3, random_state=1).fit(data)
```

Comment:
This is weird here since `data` is not dense. I would expect instead:

```python
rp_dense = RandomProjection(n_components=3, random_state=1).fit(data.toarray())
rp_spase = RandomProjection(n_components=3, random_state=1).fit(data)
```

This would be inline with what the name of the variables meant. If this is the case, then we don't need to have both CSR and COO but only one of them.

Reply:
`data` here is dense.

This is because `make_sparse_random_data` returns the data in two types (dense and sparse in csr format). And from all the 9 tests where it is used, only `test_inverse_transform` uses the sparse data option, too.

This part in `make_sparse_random_data`:
```python
data_coo = sp.coo_martix(a,(b,c), shape=(n_samples, n_features))
return data_coo.toarray(), data_coo.tocsr()
```
I wonder why it turns a coo_matrix into a scr_matrix. Is there a reason for this? If not, can I simplify the function?

I'd also like to rename `data, _ = make_sparse_random_data(...)`
into something like what was used in `test_inverse_transform`:
 `X_dense, X_csr = make_sparse_random_data(...)`
It's clearer to read.

What do you think, @glemaitre ?

Reply:
> Another option (but I don't know how much refactoring it requires) is to always return only a single type. Instead of coo_container, we can have a sparse_format: if None, we return a dense matrix, otherwise we pass the type of container that we desired. A bit similar to the changes in https://github.com/scikit-learn/scikit-learn/pull/27438.

@glemaitre 
I will push a proposal in a minute, but it doesn't get rid of `coo_container` as a param. I have tried to substitute `coo_container` with `sparse_format`, but now I believe that we have to keep `coo_container`, because we need to parametrize those datatypes in all of the tests that use `make_sparse_random_data`. Please let me know if this makes sense.

Reply:
> Ah, and there's a function def make_sparse_random_data that is exactly equal in benchmarks/bench_random_projections.py. Should this one be changed the same way?

You can let the benchmark on its own.